app/services/hashtag_autocomplete_service.rb

```diff
@@ -181,15 +201,71 @@ def search(term, types_in_priority_order, limit: 5)
     # For example, if there is a category with the slug #general and a tag
     # with the slug #general, then the tag will have its ref changed to #general::tag
     top_ranked_type = slugs_by_type.keys.first
-    results.each do |hashtag_item|
+    limited_results.each do |hashtag_item|
       next if hashtag_item.type == top_ranked_type
 
-      other_slugs = results.reject { |r| r.type === hashtag_item.type }.map(&:slug)
+      other_slugs = limited_results.reject { |r| r.type === hashtag_item.type }.map(&:slug)
       if other_slugs.include?(hashtag_item.slug)
         hashtag_item.ref = "#{hashtag_item.slug}::#{hashtag_item.type}"
       end
     end
 
-    results.take(limit)
+    limited_results.take(limit)
+  end
+
+  # TODO (martin) Remove this once plugins are not relying on the old lookup
+  # behavior via HashtagsController when enable_experimental_hashtag_autocomplete is removed
+  def lookup_old(slugs)
+    raise Discourse::InvalidParameters.new(:slugs) if !slugs.is_a?(Array)
+
+    all_slugs = []
+    tag_slugs = []
+
+    slugs[0..HashtagAutocompleteService::HASHTAGS_PER_REQUEST].each do |slug|
+      if slug.end_with?(PrettyText::Helpers::TAG_HASHTAG_POSTFIX)
+        tag_slugs << slug.chomp(PrettyText::Helpers::TAG_HASHTAG_POSTFIX)
+      else
+        all_slugs << slug
+      end
+    end
+
+    # Try to resolve hashtags as categories first
+    category_slugs_and_ids =
+      all_slugs.map { |slug| [slug, Category.query_from_hashtag_slug(slug)&.id] }.to_h
+    category_ids_and_urls =
+      Category
+        .secured(guardian)
+        .select(:id, :slug, :parent_category_id) # fields required for generating category URL
+        .where(id: category_slugs_and_ids.values)
+        .map { |c| [c.id, c.url] }
+        .to_h
+    categories_hashtags = {}
+    category_slugs_and_ids.each do |slug, id|
+      if category_url = category_ids_and_urls[id]
+        categories_hashtags[slug] = category_url
+      end
+    end
+
+    # Resolve remaining hashtags as tags
+    tag_hashtags = {}
+    if SiteSetting.tagging_enabled
+      tag_slugs += (all_slugs - categories_hashtags.keys)
+      DiscourseTagging
+        .filter_visible(Tag.where_name(tag_slugs), guardian)
+        .each { |tag| tag_hashtags[tag.name] = tag.full_url }
+    end
+
+    { categories: categories_hashtags, tags: tag_hashtags }
+  end
+
+  private
+
+  # Somtimes a specific ref is required, e.g. for categories that have
+  # a parent their ref will be parent_slug:child_slug, though most of the
+  # time it will be the same as the slug. The ref can then be used for
+  # lookup in the UI.
+  def set_refs(hashtag_items)
+    return hashtag_items if hashtag_items.blank? || hashtag_items.empty?
```

Comment:
isn't `blank?` enough here? what is testing `empty?`

Reply:
You are right, not needed